src/APNS.js

```diff
@@ -104,13 +104,14 @@ APNS.prototype.send = function(data, devices) {
   let coreData = data.data;
   let expirationTime = data['expiration_time'];
   let notification = generateNotification(coreData, expirationTime);
+  let allPromises = [];
 
-  let promises = devices.map((device) => {
+  // Start by clustering the devices per connections
+  let devicesPerConnIndex = devices.reduce((memo, device) => {
```

Comment:
I'll be honest, I don't really like this usage of `reduce`. `reduce` is from a functional paradigm, but when you modify `memo` inside your `reduce`, that feels very imperative. I'd rather see imperative code use imperative paradigms. If you wanted to copy `memo` with a modification, that would feel better, I think. Or just use a for loop. Or go back to `map`, and have the mapped function return a `(connIndex, device)` pair, since you just end up iterating over the (connIndex, device) pairs in arbitrary order anyway.